### PR TITLE
Depth First Search in Go

### PR DESCRIPTION
**Pull request by m-code12**
Fixes #2172 

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
- Added Depth First Search in Go

#### Languages Used:
- Go

#### Files Added:
- Depth_First_Search.go

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

**Comment by m-code12** [comment]
I have implemented Depth First Search in Go following all the required coding conventions. Please review and merge @jainaman224 @singh-shreya6 @asha15 @raghav-dalmia @somya-kapoor

File: Depth_First_Search/Depth_First_Search.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    var g = Graph{}
+    g.adjList = make(map[int]set)
+    return &g
+}
+
+// Add an edge between the two given nodes.If a node does not exist in the graph yet, add it.
+
+func (g *Graph) addEdge(node1 int, node2 int) {
+  
+  // Check if node1 is already in the graph
```
Comment (singh-shreya6) [inline Depth_First_Search/Depth_First_Search.go:27]: Hey, you don't need to add a blank line at all places. You can give a blank line after one if block

File: Depth_First_Search/Depth_First_Search.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    
+        g.adjList[node1] = make(set)
+    }
+	
+    g.adjList[node1][node2] = true
+
+    // Check if node2 is already in the graph
+	
+    if g.adjList[node2] == nil {
+
```
Comment (singh-shreya6) [inline Depth_First_Search/Depth_First_Search.go:39]: Please remove unnecessary blank lines.

File: Depth_First_Search/Depth_First_Search.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	
+    g.adjList[node2][node1] = true
+}
+
+// Perform a depth first search from the given node and apply the given function to each node.
+
+func dfs(g *Graph, current int, visited set, visitFunction func(int)) {
+    
+	if _, seen := visited[current]; seen {
+
```
Comment (singh-shreya6) [inline Depth_First_Search/Depth_First_Search.go:51]: Please remove unnecessary blank line

File: Depth_First_Search/Depth_First_Search.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+	if _, seen := visited[current]; seen {
+    
+	    return
+    }
+
+    visited[current] = true
+    visitFunction(current)
+
+    for neighbour := range g.adjList[current] {
+
```
Comment (singh-shreya6) [inline Depth_First_Search/Depth_First_Search.go:59]: Please remove blank line.

**Comment by m-code12** [comment]
@singh-shreya6 I have made the respective changes . Kindly review the changes .

File: Depth_First_Search/Depth_First_Search.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+type set map[int]bool
+
+// Define Graph structure
+type Graph struct {
+    adjList map[int]set
+}
+
+// Create a graph and initialize its adjacency list, and return a pointer to it.
+func createGraph() *Graph {
+      var g = Graph{}
```
Comment (singh-shreya6) [inline Depth_First_Search/Depth_First_Search.go:12]: Please ensure correct indentation. 4 spaces.

File: Depth_First_Search/Depth_First_Search.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    }
+}
+
+func main() {
+    graph := createGraph()
+    var n int
+    fmt.Println("Enter the number of vertices : ")
+    fmt.Scan(&n)
+    for i := 0; i < n; i++ {
+	    var a int
```
Comment (singh-shreya6) [inline Depth_First_Search/Depth_First_Search.go:52]: Indentation is more than 4 spaces here.

File: Depth_First_Search/Depth_First_Search.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+// Perform a depth first search from the given node and apply the given function to each node.
+func dfs(g *Graph, current int, visited set, visitFunction func(int)) {
+    if _, seen := visited[current]; seen {
+        return
+    }
+
+    visited[current] = true
+    visitFunction(current)
+	for neighbour := range g.adjList[current] {
```
Comment (singh-shreya6) [inline Depth_First_Search/Depth_First_Search.go:41]: Incorrect indentation. It should be in line with the above statement.

File: Depth_First_Search/Depth_First_Search.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+func (g *Graph) addEdge(node1 int, node2 int) {
+    // Check if node1 is already in the graph
+    if g.adjList[node1] == nil {
+        g.adjList[node1] = make(set)
+    }
+	
+    g.adjList[node1][node2] = true
+    // Check if node2 is already in the graph
+	if g.adjList[node2] == nil {
+        g.adjList[node2] = make(set)
```
Comment (singh-shreya6) [inline Depth_First_Search/Depth_First_Search.go:27]: Is it 4 spaces here after if block?

**Comment by m-code12** [comment]
When I open the same file in my pc,all identations are showing correct but here there is incorrect indentation .What should I do?

**Comment by m-code12** [comment]
I have corrected the indentations. Please review @singh-shreya6

File: Depth_First_Search/Depth_First_Search.go [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    }
+}
+
+func main() {
+    graph := createGraph()
+    var n int
+    fmt.Println("Enter the number of vertices : ")
+    fmt.Scan(&n)
+    for i := 0; i < n; i++ {
+	    var a int
```
Comment (singh-shreya6) [inline Depth_First_Search/Depth_First_Search.go:52]: There is some problem with indentation here. It is more than 4 spaces.

File: Depth_First_Search/Depth_First_Search.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+	    var b int
+	    fmt.Println("Enter the source vertex : ")
+	    fmt.Scan(&a)
+	    fmt.Println("Enter the destination vertex : ")
+	    fmt.Scan(&b)
+	    graph.addEdge(a, b)
+	}
```
Comment (singh-shreya6) [inline Depth_First_Search/Depth_First_Search.go:59]: This bracket should be in line with for statement.

**Comment by m-code12** [comment]
When I open the file in my pc , it is showing correct indentation for the statements and bracket .
But here, its incorrect .
What should I do ?

**Comment by m-code12** [comment]
Made the changes . Please review @singh-shreya6

**Pull request merged**